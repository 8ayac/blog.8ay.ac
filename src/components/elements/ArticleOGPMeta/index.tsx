import { config } from '@/site.config';
import { getLastModifiedDate } from '@/src/shared/utils';
import { Article } from '@/src/types';
import React from 'react';

export const ArticleOGPMeta: React.FC<{
  article: Article;
}> = ({ article }) => (
  <>
    <meta
      property="article:published_time"
      content={new Date(article.attributes.publishedAt).toISOString()}
    />
    <meta
      property="article:modified_time"
      content={getLastModifiedDate(article).toISOString()}
    />
    <meta property="article:author" content={config.site.maintainer} />
    {[...new Set(article.attributes.tags)].map((tag) => (
      <meta key={tag} property="article:tag" content={tag} />
    ))}
  </>
);
