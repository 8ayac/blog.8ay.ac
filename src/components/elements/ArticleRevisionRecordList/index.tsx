import { REPOSITORY_URL } from '@/src/constants/site';
import { ArticleChangeLog } from '@/src/types';
import styled from '@emotion/styled';
import React from 'react';
import urljoin from 'url-join';

const S = {
  Details: styled.details``,

  Summary: styled.summary`
    margin-bottom: 0.75rem;
    font-weight: 600;
    cursor: pointer;
    outline: none;
  `,

  LogList: styled.ul`
    padding-left: 0;
    margin: 0;
  `,

  LogListItem: styled.li`
    margin-bottom: 0.75rem;
    list-style: none;
  `,

  UpdateDateTime: styled.time``,

  UpdateDescriptionSpan: styled.span`
    font-weight: bolder;
  `,

  ToGitHubCommitAnchor: styled.a``,
};

const removeGitCommitPrefix = (commitMessage: string): string => {
  return commitMessage.replace(/^:.*?:\s+/, '');
};

const getGitCommitUrl = (commitId: string): string => {
  return urljoin(REPOSITORY_URL, 'commit', commitId);
};

const LogList: React.FC<{ log: ArticleChangeLog[] }> = ({ log }) => {
  return (
    <S.LogList>
      {log.map((record) => (
        <S.LogListItem key={record.id}>
          <S.UpdateDateTime>
            {new Date(record.date).toISOString()}
          </S.UpdateDateTime>{' '}
          <S.UpdateDescriptionSpan>
            {removeGitCommitPrefix(record.description)}
          </S.UpdateDescriptionSpan>{' '}
          (
          <S.ToGitHubCommitAnchor
            href={getGitCommitUrl(record.id)}
            target="_blank"
            rel="noopener"
          >
            #{record.id}
          </S.ToGitHubCommitAnchor>{' '}
          by {record.author})
        </S.LogListItem>
      ))}
    </S.LogList>
  );
};

export const ArticleRevisionRecordList: React.FC<{
  log: ArticleChangeLog[];
}> = ({ log }) => {
  log.sort((a, b) => new Date(a.date).getTime() - new Date(b.date).getTime());

  if (log.length < 1) return <></>;

  return (
    <>
      <S.Details open>
        <S.Summary>Revision History</S.Summary>
        <LogList log={log} />
      </S.Details>
    </>
  );
};
