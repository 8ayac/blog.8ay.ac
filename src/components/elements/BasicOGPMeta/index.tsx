import { config } from '@/site.config';
import React from 'react';
import urljoin from 'url-join';

export const BasicOGPMeta: React.FC<{
  ogTitle: string;
  ogType: string;
  ogDescription: string;
  ogImageUrl?: string;
  ogUrl: string;
}> = ({ ogTitle, ogDescription, ogImageUrl, ogUrl, ogType }) => (
  <>
    <meta property="og:title" content={ogTitle} />
    <meta property="og:type" content={ogType} />
    <meta property="og:url" content={ogUrl} />
    <meta
      property="og:image"
      content={ogImageUrl || urljoin(config.site.rootUrl, 'og.png')}
    />
    <meta property="og:site_name" content={config.site.title} />
    {!!ogDescription && (
      <meta property="og:description" content={ogDescription} />
    )}
  </>
);
