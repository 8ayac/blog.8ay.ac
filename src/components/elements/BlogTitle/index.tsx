import styled from '@emotion/styled';
import React from 'react';

const S = {
  ComponentWrapperDiv: styled.div`
    user-select: none;
  `,

  TitleSpan: styled.span`
    font-size: 2.2rem;
  `,

  SubtitleSpan: styled.span`
    margin-left: 0.5rem;
    font-size: 1.4rem;
  `,
};

export const BlogTitle: React.FC<{ title: string; subtitle: string }> = ({
  title,
  subtitle,
}) => (
  <>
    <S.ComponentWrapperDiv>
      <S.TitleSpan>{title}</S.TitleSpan>
      <S.SubtitleSpan>{subtitle}</S.SubtitleSpan>
    </S.ComponentWrapperDiv>
  </>
);
