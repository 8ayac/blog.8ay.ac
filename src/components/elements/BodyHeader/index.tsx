import { theme } from '@/src/constants/theme';
import styled from '@emotion/styled';
import Link from 'next/link';
import React, { ReactNode } from 'react';

export const S = {
  WrapperHeader: styled.header`
    margin-bottom: 3rem;
    border-bottom: 0.25rem solid ${theme.color.border.primaryLight};
  `,

  TitleH1: styled.h1`
    margin: 0 0 1rem;
    font-size: 2.16em;
  `,

  TitleAnchor: styled.a`
    margin: 0;
    color: ${theme.color.text.primary};
    text-decoration: none;
    filter: drop-shadow(2px 2px 1px ${theme.color.green.light2});
  `,

  BeforeTitleDiv: styled.div`
    margin-bottom: 0.3rem;
  `,

  AfterTitleDiv: styled.div`
    margin-top: 0.3rem;
  `,
};

export const BodyHeader: React.FC<{
  title: string;
  linkTo?: string;
  beforeTitle?: ReactNode;
  afterTitle?: ReactNode;
}> = ({ title, linkTo, beforeTitle, afterTitle }) => (
  <>
    <S.WrapperHeader>
      {beforeTitle && <S.BeforeTitleDiv>{beforeTitle}</S.BeforeTitleDiv>}
      {linkTo ? (
        <S.TitleH1>
          <Link href={linkTo} passHref>
            <S.TitleAnchor>{title}</S.TitleAnchor>
          </Link>
        </S.TitleH1>
      ) : (
        <S.TitleH1>{title}</S.TitleH1>
      )}
      {afterTitle && <S.AfterTitleDiv>{afterTitle}</S.AfterTitleDiv>}
    </S.WrapperHeader>
  </>
);
