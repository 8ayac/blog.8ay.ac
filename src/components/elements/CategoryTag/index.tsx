import { theme } from '@/src/constants/theme';
import { getTagsPagePath } from '@/src/shared/utils';
import styled from '@emotion/styled';
import Link from 'next/link';
import React from 'react';

const S = {
  Anchor: styled.a`
    padding: 3% 1rem 4%;
    margin: 0;
    font-size: 0.8em;
    font-weight: 600;
    color: white;
    text-decoration: none;
    background-color: ${theme.color.blue.dark};
    border-radius: 3px;

    &:hover {
      color: white;
      background-color: ${theme.color.blue.base};
    }
  `,
};

export const CategoryTag: React.FC<{ name: string }> = ({ name }) => (
  <>
    <Link href={getTagsPagePath(name)} passHref>
      <S.Anchor>{name}</S.Anchor>
    </Link>
  </>
);
