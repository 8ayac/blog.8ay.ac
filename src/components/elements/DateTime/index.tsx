import styled from '@emotion/styled';
import React from 'react';

const S = {
  ComponentWrapperDiv: styled.div`
    display: flex;
    flex-flow: row wrap;
  `,

  DescriptionSpan: styled.span`
    margin-right: 0.5rem;
    font-weight: 600;
  `,

  ISODateTime: styled.time`
    font-weight: bolder;
    color: ${(props) => props.theme.color.dateTime.isoDateTime.text};
  `,
};

export const DateTime: React.FC<{
  date: Date | string;
  description?: string;
}> = ({ date, description }) => (
  <>
    <S.ComponentWrapperDiv>
      {description && (
        <S.DescriptionSpan>{`${description}:`}</S.DescriptionSpan>
      )}
      <S.ISODateTime>{new Date(date).toISOString()}</S.ISODateTime>
    </S.ComponentWrapperDiv>
  </>
);
