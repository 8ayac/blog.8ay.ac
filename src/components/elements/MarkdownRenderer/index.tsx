import { theme } from '@/src/constants/theme';
import styled from '@emotion/styled';
import Prism from 'prismjs';
import React, { useEffect } from 'react';
import ReactMarkdown from 'react-markdown';
import remarkGfm from 'remark-gfm';

const S = {
  ConvertedMdBody: styled.div`
    font-size: 1.6rem;
    overflow-wrap: break-word;

    h1,
    h2 {
      padding-bottom: 0.3em;
      margin-top: 2em;
      border-bottom: 1px solid ${theme.color.border.primaryLight};
    }

    h3,
    h4,
    h5,
    h6 {
      margin-top: 1em;
    }

    a {
      font-weight: 600;
    }

    code,
    pre {
      font-family: SFMono-Regular, Consolas, Liberation Mono, Menlo, monospace;
      word-break: normal;
    }

    pre {
      white-space: pre-wrap;
    }

    /* This rule is for the inline code block. */
    code:not([class]) {
      padding: 0.2em 0.4em;
      margin: 0;
      font-size: 0.85em;
      background-color: hsla(0, 0%, 92.5%);
      border-radius: 3px;
    }

    blockquote,
    details,
    dl,
    ol,
    p,
    pre,
    table,
    ul {
      margin-top: 0;
      margin-bottom: 1.8rem;
    }

    ol ol,
    ul ol {
      list-style-type: lower-roman;
    }

    li {
      > p {
        margin-top: 1.6rem;
      }

      + li {
        margin-top: 0.25em;
      }
    }

    blockquote {
      padding: 0 1em;
      margin-left: 0;
      color: ${theme.color.text.primaryLight};
      border-left: 0.25em solid ${theme.color.blue.light3};

      :first-of-type {
        margin-top: 0;
      }

      :last-of-type {
        margin-bottom: 0;
      }
    }

    /* Below is for the plugin of Prism 'Show Language' */
    div.code-toolbar > .toolbar {
      top: 0.3em;
      right: 0.3em;
      transition: none;

      span {
        font-size: 0.85em;
        border-radius: 10px;
      }
    }
  `,
};

export const MarkdownRender: React.FC<{ content: string }> = ({ content }) => {
  useEffect(() => {
    Prism.highlightAll();
  });

  return (
    <>
      <S.ConvertedMdBody>
        <ReactMarkdown source={content} plugins={[remarkGfm]} />
      </S.ConvertedMdBody>
    </>
  );
};
