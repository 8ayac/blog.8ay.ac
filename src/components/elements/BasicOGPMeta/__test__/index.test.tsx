import { config } from '@/site.config';
import { BasicOGPMeta } from '@/src/components/elements/BasicOGPMeta';
import { mount } from 'enzyme';
import React from 'react';
import urljoin from 'url-join';

describe('BasicOGPMeta', () => {
  const testProps = {
    ogTitle: 'Test Title',
    ogType: 'testtype',
    ogDescription: 'This is a test Description',
    ogImageUrl: 'https://blog.8ay.ac/favicon.png',
    ogUrl: 'https://blog.8ay.ac/THIS_IS_TEST',
  };

  test('is rendered correctly to match the snapshot', () => {
    const wrapper = mount(<BasicOGPMeta {...testProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('has correct metadata', () => {
    const wrapper = mount(<BasicOGPMeta {...testProps} />);

    test('og:title', () => {
      expect(wrapper.find('meta[property="og:title"]').prop('content')).toBe(
        'Test Title'
      );
    });

    test('og:type', () => {
      expect(wrapper.find('meta[property="og:type"]').prop('content')).toBe(
        'testtype'
      );
    });

    test('og:description', () => {
      expect(
        wrapper.find('meta[property="og:description"]').prop('content')
      ).toBe('This is a test Description');
    });

    test('og:image', () => {
      expect(wrapper.find('meta[property="og:image"]').prop('content')).toBe(
        'https://blog.8ay.ac/favicon.png'
      );
    });

    test('og:url', () => {
      expect(wrapper.find('meta[property="og:url"]').prop('content')).toBe(
        'https://blog.8ay.ac/THIS_IS_TEST'
      );
    });

    test('og:site_name', () => {
      expect(
        wrapper.find('meta[property="og:site_name"]').prop('content')
      ).toBe(config.site.title);
    });
  });

  it('set og:image content to default value when ogImage is null', () => {
    const wrapper = mount(
      <BasicOGPMeta {...testProps} ogImageUrl={undefined} />
    );
    expect(wrapper.find('meta[property="og:image"]').prop('content')).toBe(
      urljoin(config.site.rootUrl, 'og.png')
    );
  });
});
