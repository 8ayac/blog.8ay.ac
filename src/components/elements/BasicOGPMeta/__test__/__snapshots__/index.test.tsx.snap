// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicOGPMeta is rendered correctly to match the snapshot 1`] = `
<BasicOGPMeta
  ogDescription="This is a test Description"
  ogImageUrl="https://blog.8ay.ac/favicon.png"
  ogTitle="Test Title"
  ogType="testtype"
  ogUrl="https://blog.8ay.ac/THIS_IS_TEST"
>
  <meta
    content="Test Title"
    property="og:title"
  />
  <meta
    content="testtype"
    property="og:type"
  />
  <meta
    content="https://blog.8ay.ac/THIS_IS_TEST"
    property="og:url"
  />
  <meta
    content="https://blog.8ay.ac/favicon.png"
    property="og:image"
  />
  <meta
    content="もしくはこれ"
    property="og:site_name"
  />
  <meta
    content="This is a test Description"
    property="og:description"
  />
</BasicOGPMeta>
`;
